server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -1285,4 +1304,153 @@ public BlockCacheConfiguration getBlockCacheConfiguration(AccumuloConfiguration
+      log.warn("Unloading onDemand tablet: {} for table: {} due to low memory", oldestKeyExtent,
+          oldestKeyExtent.tableId());
+      getContext().getAmple().mutateTablet(oldestKeyExtent).deleteOnDemand().mutate();
+      onDemandUnloadedLowMemory.addAndGet(1);
+      return;
+    }
+
+    // onDemandTabletAccessTimes is a HashMap. Sort the extents
+    // so that we can process them by table.
+    final SortedMap<KeyExtent,AtomicLong> sortedOnDemandExtents =
+        new TreeMap<KeyExtent,AtomicLong>();
+    sortedOnDemandExtents.putAll(onDemandTabletAccessTimes);
+
+    // The access times are updated when getOnlineTablet is called by other methods,
+    // but may not necessarily capture whether or not the Tablet is currently being used.
+    // For example, getOnlineTablet is called from startScan but not from continueScan.
+    // Instead of instrumenting all of the locations where the tablet is touched we
+    // can use the Tablet metrics.
+    final Set<KeyExtent> onDemandTabletsInUse = new HashSet<>();
+    for (KeyExtent extent : sortedOnDemandExtents.keySet()) {
+      if (isTabletInUse(extent)) {
+        onDemandTabletsInUse.add(extent);
+      }
+    }
+    if (!onDemandTabletsInUse.isEmpty()) {
+      log.debug("Removing onDemandAccessTimes for tablets as tablets are in use: {}",
+          onDemandTabletsInUse);
+      onDemandTabletsInUse.forEach(sortedOnDemandExtents::remove);
+      if (sortedOnDemandExtents.isEmpty()) {
+        return;
+      }
+    }
+
+    Set<TableId> tableIds = sortedOnDemandExtents.keySet().stream().map((k) -> {
+      return k.tableId();
+    }).distinct().collect(Collectors.toSet());
+    log.debug("Tables that have online onDemand tablets: {}", tableIds);
+    final Map<TableId,OnDemandTabletUnloader> unloaders = new HashMap<>();
+    tableIds.forEach(tid -> {
+      TableConfiguration tconf = getContext().getTableConfiguration(tid);
+      String tableContext = ClassLoaderUtil.tableContext(tconf);
+      String unloaderClassName = tconf.get(Property.TABLE_ONDEMAND_UNLOADER);
+      try {
+        Class<? extends OnDemandTabletUnloader> clazz = ClassLoaderUtil.loadClass(tableContext,
+            unloaderClassName, OnDemandTabletUnloader.class);
+        unloaders.put(tid, clazz.getConstructor().newInstance());
+      } catch (ClassNotFoundException | InstantiationException | IllegalAccessException
+          | IllegalArgumentException | InvocationTargetException | NoSuchMethodException
+          | SecurityException e) {
+        log.error(
+            "Error constructing OnDemandTabletUnloader implementation, not unloading onDemand tablets",
+            e);
+        return;
+      }
+    });
+    tableIds.forEach(tid -> {
+      Map<KeyExtent,AtomicLong> subset = sortedOnDemandExtents.entrySet().stream().filter((e) -> {
+        return e.getKey().tableId().equals(tid);
+      }).collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
```

Comment:
```suggestion
          AtomicLong> subset = sortedOnDemandExtents.entrySet().stream()
              .filter((e) -> e.getKey().tableId().equals(tid))
              .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
```

Reply:
This is just stylistic so don't need to change it if you don't want

Reply:
Addressed in 85d34f9